### ISSUE TITLE
Allow dummy function to be redefined

### ISSUE THREAD

**Issue by eight04** [PyCQA/pylint#1369]
Some services allow us to use decorator to register callbacks, to avoid creating different function and passing them into the service. Like `functools.singledispatch`:
```python
from functools import singledispatch

@singledispatch
def func(arg):
    pass

@func.register(int)
def _(arg):
    pass

@func.register(str)
def _(arg):
    pass
```
This has been addressed in #1032.

### Steps to reproduce
However, it doesn't work if we are creating our own registry:
```python
callbacks = {}

def register(name):
	def inner(func):
		callbacks[name] = func
	return inner

@register("a")
def dummy(arg):
    pass

@register("b")
def dummy(arg):
    pass
```

### Current behavior
```
E: 13, 0: function already defined line 9 (function-redefined)
```

### Expected behavior
No function-redefined warning. Probably just allow functions matching "dummy-variables-rgx" to be redefined.

### pylint --version output
```
No config file found, using default configuration
pylint 1.7.0,
astroid 1.5.0
Python 3.6.0 (v3.6.0:41df79263a11, Dec 23 2016, 07:18:10) [MSC v.1900 32 bit (In
tel)]
```

P.S. It doesn't work without decorator either, but I guess this is not a valid use case.
```python
from functools import singledispatch

@singledispatch
def func(arg):
    pass

def _(arg):
    pass
func.register(int)(_)

def _(arg):
    pass
func.register(str)(_)
```

**Comment by rogalski** [comment]
It just needs some extra ifs in [`_check_redefinition`](https://github.com/PyCQA/pylint/blob/3ccda499db433200e783d5be76004930817efe47/pylint/checkers/base.py#L604) and some test coverage. Hopefully some less experienced contributor can take on that :) .
